src/ronin/basefee-treasury/RoninBaseFeeTreasury.sol

```diff
@@ -0,0 +1,545 @@
+import { LibArray } from "src/libraries/LibArray.sol";
+
+contract RoninBaseFeeTreasury is EIP712, Initializable, HasContracts, GlobalConfigConsumer, IBaseFeeTreasury {
+  using LibArray for *;
+  using SafeCast for uint256;
+  using ErrorHandler for bool;
+  using EnumerableSet for EnumerableSet.Bytes32Set;
+
+  uint256 public constant COOLDOWN_PERIOD = 1 days;
+  uint256 public constant MIN_PROPOSAL_DURATION = 1 days;
+  bytes32 internal constant _TYPE_HASH = keccak256(
+    "Proposal(address proposer,uint32 nonce,uint40 expiry,address executor,address[] recipients,uint96[] amounts,bytes[] callDatas)"
+  );
+
+  uint256[50] private __gap;
+
+  /// @dev Global nonce for proposals.
+  uint32 internal _globalNonce;
+  /// @dev Threshold for passing proposals.
+  Threshold internal _threshold;
+  /// @dev Pending proposal set.
+  EnumerableSet.Bytes32Set internal _pendingProposals;
+  /// @dev Mapping from hash to proposal info.
+  mapping(bytes32 hash => ProposalInfo) internal _info;
+
+  modifier onlyCandidateAdmin(
+    address cid
+  ) {
+    _requireCandidateAdmin(cid);
+    _;
+  }
+
+  modifier onlyState(bytes32 hash, State state) {
+    _requireState(hash, state);
+    _;
+  }
+
+  modifier syncPendingProposals() {
+    _syncPendingProposals();
+    _;
+  }
+
+  constructor() EIP712(type(RoninBaseFeeTreasury).name, "1") {
+    _disableInitializers();
+  }
+
+  function initialize(
+    address profile,
+    address staking,
+    address validatorSet,
+    uint8 num,
+    uint8 denom
+  ) external initializer {
+    _setThreshold(num, denom);
+    _setContract(ContractType.PROFILE, profile);
+    _setContract(ContractType.STAKING, staking);
+    _setContract(ContractType.VALIDATOR, validatorSet);
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function setThreshold(uint8 num, uint8 denom) external onlyAdmin syncPendingProposals {
+    _setThreshold(num, denom);
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function incrementNonce() external onlyAdmin syncPendingProposals {
+    _incrementNonce();
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function vote(
+    address cid,
+    bytes32 hash,
+    Vote v
+  ) external onlyCandidateAdmin(cid) onlyState(hash, State.Active) syncPendingProposals {
+    _vote(hash, cid, v);
+    _tryExecute(hash);
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function propose(
+    Proposal calldata proposal
+  ) external onlyCandidateAdmin(proposal.proposer) syncPendingProposals returns (bytes32 hash) {
+    validateProposal(proposal);
+
+    hash = hashProposal(proposal);
+
+    _requireState(hash, State.Unknown);
+
+    _createProposal(hash, proposal, _snapshotVotePower(hash));
+    _vote(hash, proposal.proposer, Vote.For);
+    _tryExecute(hash);
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function cancel(
+    bytes32 hash
+  ) external onlyCandidateAdmin(_info[hash]._proposal.proposer) onlyState(hash, State.Active) syncPendingProposals {
```

Comment:
If a proposal has passed but hasn’t been executed yet, it can’t be canceled. Is that expected?

Reply:
Expected.